### ISSUE TITLE
up vs downregulated genes in a geneset

### ISSUE THREAD

**Issue by redst4r** [Gibbsdavidl/gssnng#18]
the geneset-class supports sets where some genes are positively associated with a signature (UP) and other that are negatively associated (DN).

Do we even utilize that in the scoring? 

One way (seen in some papers):
- calculate a score `s_up` for positive genes in the set
- calculate a score `s_dn` for negative  genes in the set
- the final score is their difference `s_up-s_dn`

**Comment by Gibbsdavidl** [comment]
Yeah, there’s some gene sets that are defined as being ‘down’ and some are
defined in two parts, one up, one down.  Currently, if you have a gene set
defined as DN, and the genes are low ranked, then the score is high.

So in the Hallmark genesets at MSigDB, there’s a UV response gene set
that’s defined in two parts, UP and DN… and it was my goal to make the
package compatible with MSigDB.

https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/HALLMARK_UV_RESPONSE_DN.html


But in the future, might not be important, but there’s a use case anyway.

-dave

**Comment by redst4r** [comment]
it's definitely a good thing to account for, a bunch of those "newer" genesets are weighted (some with negative weights), e.g. [Progeny](https://www.nature.com/articles/s41467-017-02391-6)

I just wasn't sure if the scoring functions handle it accordingly. It seems to be done 
in [score_fun](https://github.com/Gibbsdavidl/gssnng/blob/9ec8a43f19a30b52196fef4e6f2ffd3d72891f5c/gssnng/score_funs.py#LL258C1-L259C1). 

This is for the non-ranked version
```python
        if (gs.mode == 'UP') and (ranked == False):
            res0 = method_selector(gs, x, 'counts', gs.genes_up, method, method_params)

        elif (gs.mode == 'DN') and (ranked == False):
            res0 = method_selector(gs, x, 'counts', gs.genes_dn, method, method_params)

        elif (gs.mode == 'BOTH') and (ranked == False):
            res0_up = method_selector(gs, x, 'counts', gs.genes_up, method, method_params)
            res0_dn = method_selector(gs, x, 'counts', gs.genes_dn, method, method_params)
            res0 = (res0_up + res0_dn)
```
Here, it looks like `UP` and `DN` get handled the same way, whereas I'd have expected the "sign of `DN` to be flipped", i.e a cell with high experssion of a `DN` gene gets a low score.
Similar for `BOTH`!

For `ranked=True`:
```python 
        elif (gs.mode == 'UP') and (ranked == True):
            res0 = method_selector(gs, x, 'uprank', gs.genes_up, method, method_params)

        elif (gs.mode == 'DN') and (ranked == True):
            res0 = method_selector(gs, x, 'dnrank', gs.genes_dn, method, method_params)

        elif (gs.mode == 'BOTH') and (ranked == True):
            res0_up = method_selector(gs, x, 'uprank', gs.genes_up , method, method_params)
            res0_dn = method_selector(gs, x, 'dnrank', gs.genes_dn, method, method_params)
            res0 = (res0_up + res0_dn)
```
Looks like it's inverting the ranks for the `DN` genes by using `x['dnrank']` (didn't confirm though).
A cell with high expression of a down gene would be at the bottom of x[`dnrank`] and get a low score, which is correct.

**Comment by Gibbsdavidl** [comment]
Yeah, thanks for thinking about this.  Agree that in the ranked case we're ok, and with testing it seems to work.
However, in the counts case, maybe ` s_up-s_dn`  is a good idea!   A low score for s_dn, is what's desired with a geneset_DN.  If s_dn is high, that means the gene expression is also high, and contrary to what we want..

proposal:
'counts':  s_up - s_dn
'ranked': s_up + s_dn
